sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala

```diff
@@ -52,11 +56,13 @@ object CatalystTypeConverters {
     }
   }
 
-  private def getConverterForType(dataType: DataType): CatalystTypeConverter[Any, Any, Any] = {
+  private def getConverterForType(
+      dataType: DataType)(
+      implicit toScalaHint: Option[universe.Type] = None): CatalystTypeConverter[Any, Any, Any] = {
```

Comment:
I realize that this cuts down on the number of places that we have to pass this hint, but I think I'd prefer if we just did that explicitly.  This is a fair amount of magic, especially given that we sometimes have to extract subtypes before passing on to the next converter.